webapp/channels/src/components/button/index.tsx

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import classNames from 'classnames';
+import type {ButtonHTMLAttributes} from 'react';
+import React from 'react';
+import type {MessageDescriptor} from 'react-intl';
+import {useIntl} from 'react-intl';
+
+type ButtonProps = {
+    label: string | MessageDescriptor;
+    emphasis: 'primary' | 'secondary' | 'tertiary' | 'quaternary' | 'link';
+    size?: 'xSmall' | 'small' | 'medium' | 'large';
+    pull?: 'left' | 'right';
+    destructive?: boolean;
+    leadingIcon?: string;
+    trailingIcon?: string;
+    testId?: string;
+    loading?: boolean;
+    onClick?: ButtonHTMLAttributes<HTMLButtonElement>['onClick'];
+    autoFocus?: ButtonHTMLAttributes<HTMLButtonElement>['autoFocus'];
+    disabled?: ButtonHTMLAttributes<HTMLButtonElement>['disabled'];
+    buttonType?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
+    fullWidth?: boolean;
+}
+
+const emphasisClasses = {
+    primary: 'btn-primary',
+    secondary: 'btn-secondary',
+    tertiary: 'btn-tertiary',
+    quaternary: 'btn-quaternary',
+    link: 'btn-link',
+};
+
+const sizeClasses = {
+    xSmall: 'btn-xs',
+    small: 'btn-sm',
+    medium: '',
+    large: 'btn-lg',
+};
+
+const Button = ({
+    onClick,
+    label,
+    emphasis = 'primary',
+    size = 'medium',
+    pull,
+    destructive,
+    leadingIcon,
+    trailingIcon,
+    autoFocus,
+    testId,
+    disabled,
+    loading = false,
+    buttonType = 'button',
+    fullWidth = false,
+}: ButtonProps) => {
+    const intl = useIntl();
+    const leading = leadingIcon ? (<i className={classNames('icon', leadingIcon)}/>) : null;
+    const trailing = trailingIcon ? (<i className={classNames('icon', trailingIcon)}/>) : null;
+
+    const emphasisClass = emphasisClasses[emphasis];
+    const sizeClass = sizeClasses[size];
+
+    const labelText = typeof label === 'string' ? label : intl.formatMessage(label);
```

Comment:
Nitpicking again, but this could use `formatAsString` from `utils/i18n` that I added recently. It was only merged into master this past week though, so it wouldn't have existed when you wrote this before

Reply:
It works here, but I have to admit I am not a big fan of its potential to return undefined. If it formats as a String, it should always return a string :stuck_out_tongue: But 0/5.

Reply:
I'm trying to remember why I did that since I always wish JS was like Go where the zero value for strings is the empty string. I think it was to allow for it to have a default value.